File: pulsar-skeleton/src/main/java/ai/platon/pulsar/crawl/fetch/FetchJobForwardingResponse.java
package ai.platon.pulsar.crawl.fetch;

import ai.platon.pulsar.common.HttpHeaders;
import ai.platon.pulsar.persist.ProtocolStatus;
import ai.platon.pulsar.persist.metadata.MultiMetadata;
import ai.platon.pulsar.crawl.protocol.ForwardingResponse;
import ai.platon.pulsar.persist.metadata.ProtocolStatusCodes;
import org.apache.commons.lang3.math.NumberUtils;

import static ai.platon.pulsar.common.config.PulsarConstants.FETCH_PRIORITY_DEFAULT;

/**
 *
 * */
public class FetchJobForwardingResponse extends ForwardingResponse implements HttpHeaders {

    public FetchJobForwardingResponse(MultiMetadata headers, byte[] content) {
        super(headers.get(Q_URL), content, ProtocolStatus.fromMinor(NumberUtils.toInt(headers.get(Q_STATUS_CODE), ProtocolStatusCodes.NOTFOUND)), headers);
    }

    public int getJobId() {
        return NumberUtils.toInt(getHeaders().get(Q_JOB_ID), 0);
    }

    public int getPriority() {
        return NumberUtils.toInt(getHeaders().get(Q_PRIORITY), FETCH_PRIORITY_DEFAULT);
    }

    public String getQueueId() {
        return getHeaders().get(Q_QUEUE_ID);
    }

    public int getItemId() {
        return NumberUtils.toInt(getHeaders().get(Q_ITEM_ID), 0);
    }
}

File: pulsar-skeleton/src/main/kotlin/ai/platon/pulsar/crawl/fetch/FetchJobForwardingResponse.kt
package ai.platon.pulsar.crawl.fetch

import ai.platon.pulsar.common.HttpHeaders
import ai.platon.pulsar.common.config.AppConstants.FETCH_PRIORITY_DEFAULT
import ai.platon.pulsar.crawl.protocol.ForwardingResponse
import ai.platon.pulsar.persist.ProtocolStatus
import ai.platon.pulsar.persist.WebPage
import ai.platon.pulsar.persist.metadata.MultiMetadata
import ai.platon.pulsar.persist.metadata.ProtocolStatusCodes

class FetchJobForwardingResponse(
        headers: MultiMetadata,
        content: ByteArray
): ForwardingResponse(WebPage.newWebPage(headers.get(HttpHeaders.Q_URL)), getProtocolStatus(headers)!!, headers, content), HttpHeaders {

    val jobId: Int
        get() = headers.getInt(HttpHeaders.Q_JOB_ID, 0)

    val priority: Int
        get() = headers.getInt(HttpHeaders.Q_PRIORITY, FETCH_PRIORITY_DEFAULT)

    val queueId: String
        get() = headers.get(HttpHeaders.Q_QUEUE_ID)?:""

    val itemId: Int
        get() = headers.getInt(HttpHeaders.Q_ITEM_ID, 0)

    companion object {
        fun getProtocolStatus(headers: MultiMetadata): ProtocolStatus? {
            return ProtocolStatus.fromMinor(headers.getInt(HttpHeaders.Q_STATUS_CODE, ProtocolStatusCodes.NOTFOUND))
        }
    }
}
